### PR TITLE
Add dotnet core test app

### PR DESCRIPTION
**Pull request by gkarabin**
It would be helpful to have a test project for .NET Core.

This PR does so without materially changing the existing .NET Framework 3.5 test project.  This PR provides a new .NET Core test project, moving code that is common to both it and the original test project into shared projects.

In creating this change, I found a problem with TinyMessenger on .NET Core and fixed it. I did not attempt to deal with ASP .NET extensions, which don't compile for .NET Core.

### CODE REVIEW

**Comment by gkarabin** [comment]
Resolves #133

**Comment by gkarabin** [comment]
Some of the file moves (like to TinyIoC.cs) include whitespace changes to each line of the file. I'm not happy about that and will resolve it. I'm not sure if Visual Studio did it when I moved the files from within it, or if something else happened.

**Comment by gkarabin** [comment]
I fixed the renames without changes (used a force push, since I doubt anyone has seen this PR yet). It is in better shape now, and should be OK to merge if you can handle the structural changes to the solution.

**Comment by niemyjski** [comment]
I'm really torn here and maybe others can weigh in on this. One of the advantages of this library was a wide array of platform support and a single file, but that's starting to become irrelevant as they are no longer supported and net standard is the new king and long term successor. Which brings me to my next point, why not just support full framework (lowest still supported?) and the lowest net standard possible in one project. At what point does everyone just use the built in Microsoft DI that works on net standard (I'm all for multiple options and choice). With that said, I'd rather be simplifying and removing code than splitting this up into multiple projects. @grumpydev @benaadams @xinmyname @phillip-haydon

**Comment by gkarabin** [comment]
@niemyjski  I can remove the shared projects from the PR to reduce the number of changes. The PR would be left with two new projects (the .NET Core app and its external assembly). The .NET Core app could reference the .cs files that it needs directly.

That doesn't address your other concerns but at least it simplifies this PR.

**Comment by niemyjski** [comment]
I think we should only have one project and no shared projects as it was
before and just update things to work in the new world.
Thanks
-Blake Niemyjski

**Comment by niemyjski** [comment]
I think we should only have one project and no shared projects as it was before and just update things to work in the new world. We can always do muvtitargetting in a single project.

**Comment by gkarabin** [comment]
I updated the PR to use use multi-targeting as requested. The library projects continue to build for .NET Framework 3.5, but now TinyIoC also builds for .NET Standard 2.0 and TinyIoC.ExternalTypes also builds for .NET Standard 1.1 (1.0 did not suffice due to some attributes that were referenced).

TinyIoC.Tests now builds for .NET Framework 4.5 (rather than 4.0) and .NET Core 2.1. That was the combination that provided least-common denominator support for running tests in Visual Studio (16.5.4) with Test Explorer. I had been hoping to build .NET Framework 4.0 and .NET Core 1.0, but couldn't find packages that supported that combination (whether nugets or using 

It's not clear to me if there is a strong need to run the tests in a .NET Framework 4.0 environment or not, or if it's OK that they might bit-rot for that framework going forward.

Also note that the final commit of the PR pares down what appears to me to be unnecessary project properties, since I can get builds to work as-is on my environment. I don't have experience with the Bootstrapper properties, which were set up to ensure that the right frameworks were installed. I could have left those properties alone in the TinyIoC and TinyIoC.ExternalTypes projects (since they still target .NET Framework 3.5), but I would have had to update them in the TinyIoC.Tests project (which moved from .NET Framework 4.0 to 4.5). I didn't know how to update them for .NET 4.5. If it matters I can figure that out, but if it doesn't it seems easier to just drop those properties altogether.

File: src/TinyIoC/TinyMessenger.cs [hunk 1]
```diff
@@ -781,12 +784,23 @@ where sub.Subscription.ShouldAttemptDelivery(message)
...
 
+        #if !NETSTANDARD
         private void PublishAsyncInternal<TMessage>(TMessage message, AsyncCallback callback) where TMessage : class, ITinyMessage
         {
             Action publishAction = () => { PublishInternal<TMessage>(message); };
 
             publishAction.BeginInvoke(callback, null);
         }
+        #else
+        private async void PublishAsyncInternal<TMessage>(TMessage message, AsyncCallback callback) where TMessage : class, ITinyMessage
```
Comment (niemyjski) [inline src/TinyIoC/TinyMessenger.cs:795]: Is there any reason for adding this?

**Comment by niemyjski** [comment]
Looks good, thanks for updating that, I had a single question via comment and then I can merge it. I'm thinking that going forward we should only support what Microsoft officially supports as far as runtimes (NET 48, .NET Core 3.1 LTS), otherwise it's going to be a pain to run tests / verify code compiles on super outdated runtimes (e.g., you have to go searching for the .NET Core 2.1 runtime).

File: src/TinyIoC/TinyMessenger.cs (hunk 1 above)
Comment (gkarabin) [reply src/TinyIoC/TinyMessenger.cs:795]: Yes. I noted it in the message for the commit where I added it, but I can elaborate. Here's the commit message:

> Fix a .NET Core incompatibility with Delegate.BeginInoke in TinyMessenger (see https://devblogs.microsoft.com/dotnet/migrating-delegate-begininvoke-calls-for-net-core/).  Now all tests pass.

This is actually the change that made me want to get started on this PR in the first place. .NET Core doesn't work with BeginInvoke (the IAP model is supplanted by the TAP model) and this method wasn't working.  

So the else case works for .NET Core (or other runtimes that support .NET standard).  The link to the Microsoft devblog has all of the nitty gritty details.

**Comment by gkarabin** [comment]
I'm cool either way. 

It was maddening finding the minimum versions combination of runtimes that worked in Visual Studio today. That combination very well might not work in the future when a VS update requires a test platform update, etc.

As the code stands, there are issues running the tests outside of Visual Studio.  "dotnet test" does not work with this PR, for instance. I think that it would be more likely that it would work if all of the runtimes (and possibly the library projects) were built for newer versions.

**Comment by niemyjski** [comment]
Thanks for the pr!

**Pull request merged**